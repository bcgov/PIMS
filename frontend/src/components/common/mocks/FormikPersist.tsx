import React from 'react';

const Persist = () => <div data-testid="mocked-persist" />;

export { Persist };
