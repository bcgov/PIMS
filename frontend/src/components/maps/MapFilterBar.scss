@import '../../fonts.scss';
@import '../../variables';

.container-fluid {
  &.map-filter-container {
    height: #{$mapfilter-height};
    background-color: #f2f2f2;

    .map-filter-bar {
      align-items: center;
      padding: 1rem 0;

      .bar-item {
        .form-group {
          padding: 0;
          margin: 0;
        }
      }

      .btn-warning {
        color: white;
        font-weight: bold;
      }

      .form-control {
        border-radius: 0;
      }
    }
  }
}
