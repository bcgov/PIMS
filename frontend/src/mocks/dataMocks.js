export const createMockHeader = () => ({
  headers: {
    "Access-Control-Allow-Origin": "*",
    Authorization: "mockToken",
  },
});

export const ERROR = { message: "Errors", status: 400 };
