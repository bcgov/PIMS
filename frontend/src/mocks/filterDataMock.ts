import { ILookupCode } from '../actions/lookupActions';
import { IParcel, IProperty } from 'actions/parcelsActions';
import { MapViewportChangeEvent } from 'components/maps/leaflet/Map';

export const SELECTEDCLASSIFICATION = {
  name: 'Core Operational',
  id: 0,
  isDisabled: false,
  type: 'Classification',
};

export const DISABLED = [
  {
    name: 'Core Operational',
    id: 0,
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Core Strategic',
    id: 1,
    isDisabled: true,
    type: 'Classification',
  },
];

export const CLASSIFICATIONS = [
  {
    name: 'Core Operational',
    id: '0',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Core Strategic',
    id: '1',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Surplus Active',
    id: '2',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Surplus Encumbered',
    id: '3',
    isDisabled: false,
    type: 'Classification',
  },
] as ILookupCode[];

export const AGENCIES = [
  {
    code: 'AEST',
    name: 'AEST',
    id: '1',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'HTLH',
    name: 'HTLH',
    id: '2',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'MOTI',
    name: 'MOTI',
    id: '3',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'FLNR',
    name: 'FLNR',
    id: '4',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'MAH',
    name: 'MAH',
    id: '5',
    isDisabled: false,
    type: 'Agency',
  },
];

export const PARCELS = [
  { id: 1, latitude: 48, longitude: 123 },
  { id: 2, latitude: 50, longitude: 133 },
] as IProperty[];

export const mockDetails = [
  {
    id: 1,
    pid: '000-000-000',
    pin: '',
    projectNumber: '',
    statusId: undefined,
    municipality: '',
    zoning: '',
    zoningPotential: '',
    classificationId: undefined,
    agencyId: undefined,
    isSensitive: false,
    latitude: 48,
    longitude: 123,
    propertyStatus: 'active',
    classification: 'Core Operational',
    description: 'test',
    evaluations: [
      {
        date: new Date(),
        key: 'Assessed',
        value: 10000,
      },
    ],
    fiscals: [
      {
        fiscalYear: 2020,
        key: 'NetBook',
        value: 10000,
      },
    ],
    address: {
      line1: '1234 mock Street',
      line2: 'N/A',
      city: 'Victoria',
      province: 'BC',
      postal: 'V1V1V1',
    },
    landArea: 'unknown',
    landLegalDescription: 'test',
    buildings: [],
    agency: 'MOTI',
  },
  {
    id: 2,
    pid: '000-000-000',
    pin: '',
    statusId: undefined,
    municipality: '',
    zoning: '',
    zoningPotential: '',
    classificationId: undefined,
    agencyId: undefined,
    isSensitive: false,
    latitude: 50,
    longitude: 133,
    propertyStatus: 'active',
    classification: 'Core Operational',
    description: 'test',
    address: {
      line1: '1234 mock Street',
      line2: 'N/A',
      city: 'Victoria',
      province: 'BC',
      postal: 'V1V1V1',
    },
    landArea: 'unknown',
    landLegalDescription: 'test',
    buildings: [],
    evaluations: [
      {
        date: new Date(),
        key: 'Assessed',
        value: 10000,
      },
    ],
    fiscals: [
      {
        fiscalYear: 2020,
        key: 'NetBook',
        value: 10000,
      },
    ],
    agency: 'HLTH',
  },
] as IParcel[];

export const ACTIVE = {
  id: 1,
  pid: '000-000-000',
  pin: '',
  projectNumber: '',
  statusId: undefined,
  municipality: '',
  zoning: '',
  zoningPotential: '',
  classificationId: undefined,
  agencyId: undefined,
  isSensitive: false,
  latitude: 48,
  longitude: 123,
  propertyStatus: 'active',
  classification: 'Core Operational',
  description: 'test',
  evaluations: [
    {
      date: new Date(),
      key: 'Assessed',
      value: 10000,
    },
  ],
  fiscals: [
    {
      fiscalYear: 2020,
      key: 'NetBook',
      value: 10000,
    },
  ],
  address: {
    line1: '1234 mock Street',
    line2: 'N/A',
    city: 'Victoria',
    province: 'BC',
    postal: 'V1V1V1',
  },
  landArea: 'unknown',
  landLegalDescription: 'test',
  buildings: [],
  agency: 'FIN',
} as IParcel;

export const mockAgencyModel = {
  bounds: null,
  filter: {
    agencies: '1',
    classificationId: null,
  },
} as MapViewportChangeEvent;
