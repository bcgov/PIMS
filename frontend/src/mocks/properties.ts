import {
  IAddress,
  IBuilding,
  IEvaluation,
  IFiscal,
  IParcel,
  IProperty,
} from 'actions/parcelsActions';
import { PropertyTypes } from 'constants/index';

export const mockProperty: IProperty = {
  id: 1,
  propertyTypeId: PropertyTypes.PARCEL,
  agencyId: 1,
  agency: 'agency',
  subAgency: 'subagency',
  agencyFullName: 'agency full name',
  subAgencyFullName: 'subagency full name',
  latitude: 0.31,
  longitude: 1.34,
  name: 'name',
  description: 'description',
  projectNumbers: ['SPP-1233'],
  projectStatus: 'status',
  projectWorkflow: 'workflow',
  isSensitive: false,
  createdOn: Date.UTC.toString(),
  updatedOn: Date.UTC.toString(),
  updatedByEmail: 'john.doe@email.com',
  updatedByName: 'John Doe',
};

export const mockAddress: IAddress = {
  line1: '1234 Some Street',
  administrativeArea: 'administrative area',
  provinceId: 'BC',
  postal: 'V9V 3I3',
};

export const mockEvaluation: IEvaluation = {
  key: 'Assessed',
  value: 1000.23,
};

export const mockFiscal: IFiscal = {
  key: 'Appraised',
  value: 333.21,
};

export const mockBuilding: IBuilding = {
  ...mockProperty,
  parcelId: 1,
  pid: 12333,
  address: mockAddress,
  buildingConstructionTypeId: 1,
  buildingPredominateUseId: 1,
  buildingOccupantTypeId: 1,
  classificationId: 1,
  classification: 'classification',
  encumbranceReason: 'encumbranceReason',
  occupantName: 'occupantName',
  transferLeaseOnSale: false,
  buildingTenancy: 'buildingTenancy',
  rentableArea: 234,
  totalArea: 500,
  agencyCode: 'agencyCode',
  assessedLand: 3434.23,
  assessedBuilding: 34343.34,
  evaluations: [mockEvaluation],
  fiscals: [mockFiscal],
  parcels: [],
};

export const mockParcel: IParcel = {
  ...mockProperty,
  pid: '133-123-123',
  pin: '',
  classification: 'classification',
  classificationId: 1,
  encumbranceReason: 'encumbranceReason',
  address: mockAddress,
  landArea: 235,
  landLegalDescription: 'landLegalDescription',
  zoning: 'zoning',
  zoningPotential: 'zoningPotential',
  buildings: [mockBuilding],
  parcels: [],
  assessedLand: 12313.12,
  assessedBuilding: 2432.23,
  evaluations: [mockEvaluation],
  fiscals: [mockFiscal],
  rowVersion: 'version',
};
