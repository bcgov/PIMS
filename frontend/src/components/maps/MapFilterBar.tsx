import './MapFilterBar.scss';

import React, { useMemo } from 'react';
import { Col } from 'react-bootstrap';
import { Formik, useFormikContext } from 'formik';
import { ILookupCode } from 'actions/lookupActions';
import { Form, Select, InputGroup, Input, SelectOption } from '../common/form';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import { Claims } from 'constants/claims';
import SppButton from 'components/common/form/SppButton';
import { FilterBarSchema } from 'utils/YupSchema';
import ResetButton from 'components/common/form/ResetButton';
import SearchButton from 'components/common/form/SearchButton';
import { BasePropertyFilter } from 'components/common/interfaces';
import { ParentGroupedFilter } from 'components/SearchBar/ParentGroupedFilter';
import { mapLookupCodeWithParentString } from 'utils';

const SearchBar: React.FC = () => {
  const state: { options: any[]; placeholders: Record<string, string> } = {
    options: [
      { label: 'Address', value: 'address' },
      { label: 'Location', value: 'administrativeArea' },
      { label: 'PID/PIN', value: 'pid' },
      { label: 'RAEG or SPP No.', value: 'projectNumber' },
    ],
    placeholders: {
      address: 'Enter an address or city',
      administrativeArea: 'Enter a location name',
      pid: 'Enter a PID or PIN',
      projectNumber: 'Enter an SPP/RAEG number',
    },
  };

  // access the form context values, no need to pass props
  const {
    values: { searchBy },
    setFieldValue,
  } = useFormikContext<MapFilterChangeEvent>();
  const desc = state.placeholders[searchBy] || '';

  const reset = () => {
    setFieldValue('address', '');
    setFieldValue('administrativeArea', '');
    setFieldValue('projectNumber', '');
    setFieldValue('city', '');
  };

  return (
    <InputGroup
      fast={false}
      formikProps={null as any}
      prepend={<Select field="searchBy" options={state.options} onChange={reset} />}
      field={searchBy}
      placeholder={desc}
    ></InputGroup>
  );
};

export interface MapFilterChangeEvent extends BasePropertyFilter {
  searchBy: string;
  pid: string;
  address: string;
  administrativeArea: string;
  projectNumber: string;
  /** comma-separated list of agencies to filter by */
  agencies: string;
  classificationId: string;
  minLotSize: string;
  maxLotSize: string;
  inSurplusPropertyProgram?: boolean;
  inEnhancedReferralProcess?: boolean;
}

type MapFilterProps = {
  agencyLookupCodes: ILookupCode[];
  propertyClassifications: ILookupCode[];
  lotSizes: number[];
  mapFilter?: MapFilterChangeEvent;
  onFilterChange: (e: MapFilterChangeEvent) => void;
  onFilterReset?: () => void;
};

const defaultFilterValues: MapFilterChangeEvent = {
  searchBy: 'address',
  pid: '',
  address: '',
  administrativeArea: '',
  projectNumber: '',
  agencies: '',
  classificationId: '',
  minLotSize: '',
  maxLotSize: '',
  inSurplusPropertyProgram: false,
};

/**
 * filter overlay for the map, controls pin display.
 */
const MapFilterBar: React.FC<MapFilterProps> = ({
  agencyLookupCodes,
  propertyClassifications,
  mapFilter,
  onFilterChange,
  onFilterReset,
}) => {
  const mapLookupCode = (code: ILookupCode): SelectOption => ({
    label: code.name,
    value: code.id.toString(),
    code: code.code,
    parentId: code.parentId,
  });
  const agencies = (agencyLookupCodes ?? []).map(c =>
    mapLookupCodeWithParentString(c, agencyLookupCodes),
  );
  const classifications = (propertyClassifications ?? []).map(c => mapLookupCode(c));
  const keycloak = useKeycloakWrapper();
  let formikRef = React.useRef<any>() as any;

  const initialValues = useMemo(() => {
    const values = { ...defaultFilterValues, ...mapFilter };
    if (typeof values.agencies === 'string') {
      const agency = agencies.find(x => x.value.toString() === values.agencies.toString()) as any;
      if (agency) {
        values.agencies = agency;
      }
    }
    return values;
  }, [agencies, mapFilter]);

  const applyEnhancedReferralFilter = () => {
    const values: MapFilterChangeEvent = { ...formikRef!.values };
    values.inEnhancedReferralProcess = true;
    values.inSurplusPropertyProgram = false;
    onFilterChange(values);
  };

  const applySurplusPropertyFilter = () => {
    const values: MapFilterChangeEvent = { ...formikRef!.values };
    values.inSurplusPropertyProgram = true;
    values.inEnhancedReferralProcess = false;
    onFilterChange(values);
  };

  return (
    <Formik<MapFilterChangeEvent>
      initialValues={{ ...initialValues }}
      enableReinitialize
      validationSchema={FilterBarSchema}
      onSubmit={(values, { setSubmitting }) => {
        setSubmitting(true);
        delete values.inEnhancedReferralProcess;
        delete values.inSurplusPropertyProgram;
        onFilterChange?.({ ...values, agencies: (values.agencies as any)?.value });
        setSubmitting(false);
      }}
      innerRef={ref => (formikRef = ref)}
    >
      {({ isSubmitting, handleReset, handleSubmit, setFieldValue, values }) => (
        <Form>
          <Form.Row className="map-filter-bar align-items-start">
            <Col className="bar-item">
              <SearchBar />
            </Col>
            <Col className="bar-item">
              <ParentGroupedFilter
                name="agencies"
                options={agencies}
                className="map-filter-typeahead"
                filterBy={['code', 'label', 'parent']}
                placeholder="Enter an Agency"
                inputSize="large"
              />
            </Col>
            <Col className="bar-item">
              <Select
                field="classificationId"
                placeholder="Classification"
                options={classifications}
              />
            </Col>
            <Col className="bar-item d-flex align-items-start">
              <Input field="minLotSize" placeholder="Min Lot Size" />
              <span className="mx-2 align-self-center">-</span>
              <Input field="maxLotSize" placeholder="Max Lot Size" />
            </Col>
            {keycloak.hasClaim(Claims.ADMIN_PROPERTIES) && (
              <Col className="bar-item flex-grow-0">
                <SppButton
                  handleErpClick={applyEnhancedReferralFilter}
                  handleSppClick={applySurplusPropertyFilter}
                  inEnhancedReferralProcess={values.inEnhancedReferralProcess}
                  inSurplusPropertyProgram={values.inSurplusPropertyProgram}
                />
              </Col>
            )}
            <Col className="bar-item flex-grow-0">
              <SearchButton disabled={isSubmitting} />
            </Col>
            <Col className="bar-item flex-grow-0">
              <ResetButton
                disabled={isSubmitting}
                onClick={() => onFilterChange?.({ ...defaultFilterValues })}
              />
            </Col>
          </Form.Row>
        </Form>
      )}
    </Formik>
  );
};

export default MapFilterBar;
