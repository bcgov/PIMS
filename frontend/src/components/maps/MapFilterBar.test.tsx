import React from 'react';
import renderer from 'react-test-renderer';
import { render, wait, fireEvent } from '@testing-library/react';
import MapFilterBar, { MapFilterChangeEvent } from './MapFilterBar';
import * as MOCK from 'mocks/filterDataMock';

const onFilterChange = jest.fn<void, [MapFilterChangeEvent]>();
//prevent web calls from being made during tests.
jest.mock('axios');

describe('MapFilterBar', () => {
  it('renders correctly', () => {
    // Capture any changes
    const tree = renderer
      .create(
        <MapFilterBar
          agencyLookupCodes={MOCK.AGENCIES}
          propertyClassifications={MOCK.CLASSIFICATIONS}
          lotSizes={[1, 2, 3]}
          onFilterChange={onFilterChange}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('submits correct values', async () => {
    // Arrange
    const uiElement = (
      <MapFilterBar
        agencyLookupCodes={MOCK.AGENCIES}
        propertyClassifications={MOCK.CLASSIFICATIONS}
        lotSizes={[1, 2, 3]}
        onFilterChange={onFilterChange}
      />
    );
    const { container } = render(uiElement);
    const address = container.querySelector('input[name="address"]');
    const municipality = container.querySelector('input[name="municipality"]');
    const projectNumber = container.querySelector('input[name="projectNumber"]');
    const agencies = container.querySelector('select[name="agencies"]');
    const classificationId = container.querySelector('select[name="classificationId"]');
    const minLotSize = container.querySelector('select[name="minLotSize"]');
    const maxLotSize = container.querySelector('select[name="maxLotSize"]');
    const submit = container.querySelector('button[type="submit"]');

    // Act
    // Enter values on the form fields, then click the Search button
    await wait(() => {
      fireEvent.change(address!, {
        target: {
          value: 'mockaddress',
        },
      });
    });

    await wait(() => {
      fireEvent.change(municipality!, {
        target: {
          value: 'mockmunicipality',
        },
      });
    });

    await wait(() => {
      fireEvent.change(projectNumber!, {
        target: {
          value: 'mock-project-number',
        },
      });
    });

    await wait(() => {
      fireEvent.change(agencies!, {
        target: {
          value: '1',
        },
      });
    });

    await wait(() => {
      fireEvent.change(classificationId!, {
        target: {
          value: '0',
        },
      });
    });

    await wait(() => {
      fireEvent.change(minLotSize!, {
        target: {
          value: '1',
        },
      });
    });

    await wait(() => {
      fireEvent.change(maxLotSize!, {
        target: {
          value: '3',
        },
      });
    });

    await wait(() => {
      fireEvent.click(submit!);
    });

    // Assert
    expect(onFilterChange).toBeCalledWith<[MapFilterChangeEvent]>({
      address: 'mockaddress',
      municipality: 'mockmunicipality',
      projectNumber: 'mock-project-number',
      agencies: '1',
      classificationId: '0',
      minLotSize: '1',
      maxLotSize: '3',
    });
  });
});
