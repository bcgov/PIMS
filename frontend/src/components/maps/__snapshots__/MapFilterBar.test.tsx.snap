// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapFilterBar renders correctly 1`] = `
<div
  className="map-filter-container container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-md-8 col-8 offset-md-2 offset-2"
    >
      <form
        className=""
        noValidate={true}
        onSubmit={[Function]}
      >
        <div
          className="map-filter-bar form-row"
        >
          <div
            className="bar-item col"
          >
            <div
              className="form-group"
            >
              <input
                className="form-control"
                id="input-address"
                name="address"
                onChange={[Function]}
                placeholder="Enter an address or city"
                value=""
              />
            </div>
          </div>
          <div
            className="bar-item col"
          >
            <div
              className="form-group"
            >
              <select
                className="form-control"
                id="input-agencies"
                name="agencies"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  View Properties In
                </option>
                <option
                  value="1"
                >
                  AEST
                </option>
                <option
                  value="2"
                >
                  HTLH
                </option>
                <option
                  value="3"
                >
                  MOTI
                </option>
                <option
                  value="4"
                >
                  FLNR
                </option>
                <option
                  value="5"
                >
                  MAH
                </option>
              </select>
            </div>
          </div>
          <div
            className="bar-item col"
          >
            <div
              className="form-group"
            >
              <select
                className="form-control"
                id="input-classificationId"
                name="classificationId"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  View by Classification
                </option>
                <option
                  value="0"
                >
                  Core Operational
                </option>
                <option
                  value="1"
                >
                  Core Strategic
                </option>
                <option
                  value="2"
                >
                  Surplus Active
                </option>
                <option
                  value="3"
                >
                  Surplus Encumbered
                </option>
              </select>
            </div>
          </div>
          <div
            className="bar-item col"
          >
            <div
              className="form-group"
            >
              <select
                className="form-control"
                id="input-minLotSize"
                name="minLotSize"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  No Min Lot Size
                </option>
              </select>
            </div>
          </div>
          <div
            className="bar-item col"
          >
            <div
              className="form-group"
            >
              <select
                className="form-control"
                id="input-maxLotSize"
                name="maxLotSize"
                onChange={[Function]}
                value=""
              >
                <option
                  value=""
                >
                  No Max Lot Size
                </option>
              </select>
            </div>
          </div>
          <div
            className="bar-item flex-grow-0 col"
          >
            <button
              className="pl-0 btn btn-warning"
              disabled={false}
              type="submit"
            >
              <img
                className="mr-2"
                src="icon-search.svg"
              />
              Search
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
